Completed homework 20 sprint)

RURURuina opened the pull request:
Если что-то неправильно подскажите, как решить)))
Заранее спасибо за проверку

RURURuina committed:
Completed homework 20 sprint)
Если что-то неправильно подскажите, как решить)))
Заранее спасибо за проверку

DmitrijEgorow commented:
- [ ] Приложение аварийно завершается при смене конфигурации устройства
- [ ] Под заголовком “Вы искали” должны быть треки (макс 10 шт), на которые хоть раз кликнул пользователь, на видео заметно какое-то неконсистентное состояние, когда есть заголовок "Вы искали" и почему-то одновременно результат поискового запроса



https://github.com/user-attachments/assets/a3199351-d1eb-491c-87f7-c7daf18535c8

```
FATAL EXCEPTION: main  
Process: com.practicum.playlistmaker, PID: 25118  
java.lang.RuntimeException: Unable to start activity ComponentInfo{com.practicum.playlistmaker/com.practicum.playlistmaker.player.ui.AudioPlayerActivity}: java.lang.IllegalStateException  
at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:4169)  
at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:4325)  
at android.app.ActivityThread.handleRelaunchActivityInner(ActivityThread.java:6484)  
at android.app.ActivityThread.handleRelaunchActivity(ActivityThread.java:6341)  
at android.app.servertransaction.ActivityRelaunchItem.execute(ActivityRelaunchItem.java:71)  
at android.app.servertransaction.ActivityTransactionItem.execute(ActivityTransactionItem.java:45)  
at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:135)  
at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:95)  
at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2574)  
at android.os.Handler.dispatchMessage(Handler.java:106)  
at android.os.Looper.loopOnce(Looper.java:226)  
at android.os.Looper.loop(Looper.java:313)  
at android.app.ActivityThread.main(ActivityThread.java:8762)  
at java.lang.reflect.Method.invoke(Native Method)  
at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:604)  
at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1067)  
Caused by: java.lang.IllegalStateException  
at android.media.MediaPlayer.nativeSetDataSource(Native Method)  
at android.media.MediaPlayer.setDataSource(MediaPlayer.java:1256)  
at android.media.MediaPlayer.setDataSource(MediaPlayer.java:1243)  
at android.media.MediaPlayer.setDataSource(MediaPlayer.java:1208)  
at com.practicum.playlistmaker.player.data.impl.MediaPlayerRepositoryImpl.preparePlayer(MediaPlayerRepositoryImpl.kt:20)  
at com.practicum.playlistmaker.player.domain.impl.MediaPlayerInteractorImpl.preparePlayer(MediaPlayerInteractorImpl.kt:14)  
at com.practicum.playlistmaker.player.ui.view_model.MediaPlayerViewModel.preparePlayer(MediaPlayerViewModel.kt:52)  
at com.practicum.playlistmaker.player.ui.AudioPlayerActivity.preparePlayer(AudioPlayerActivity.kt:132)  
at com.practicum.playlistmaker.player.ui.AudioPlayerActivity.onCreate(AudioPlayerActivity.kt:118)  
at android.app.Activity.performCreate(Activity.java:8591)  
at android.app.Activity.performCreate(Activity.java:8570)  
at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1384)
```

DmitrijEgorow commented:
> * [ ]  Приложение аварийно завершается при смене конфигурации устройства
> * [ ]  Под заголовком “Вы искали” должны быть треки (макс 10 шт), на которые хоть раз кликнул пользователь, на видео заметно какое-то неконсистентное состояние, когда есть заголовок "Вы искали" и почему-то одновременно результат поискового запроса
> 
>  Screen_recording_20240824_012235.mp4

Я бы реализовал схему с набором состояний (states), а не наборов if-ов, из которых какая-то комбинация будет приводить к промежуточному неконсистентному состоянию
но можно и с другими подходами

DmitrijEgorow commented:
- [ ] Если было activity paused, то музыка не воспроизводится

https://github.com/user-attachments/assets/ed20070f-14c4-4ed1-8f61-f122ac8de8a0

DmitrijEgorow commented on app/src/main/java/com/practicum/playlistmaker/player/ui/AudioPlayerActivity.kt:
```diff
@@ -67,19 +66,16 @@ class AudioPlayerActivity : AppCompatActivity() {
         if (track.artworkUrl100?.isNotEmpty() == true) {
-            Glide.with(this)
-                .load(track.artworkUrl100.replaceAfterLast("/", "512x512bb.jpg"))
+            Glide.with(this).load(track.artworkUrl100.replaceAfterLast("/", "512x512bb.jpg"))
```
Такие строки лучше выносить в константы

RURURuina replied:
Вынес

DmitrijEgorow commented on app/src/main/java/com/practicum/playlistmaker/player/ui/view_model/MediaPlayerViewModel.kt:
```diff
@@ -1,53 +1,47 @@
 
 class MediaPlayerViewModel(private val interactor: MediaPlayerInteractor) : ViewModel() {
     private companion object {
         const val TRACK_FINISH = 29_900L
```
Не используется

RURURuina replied:
Удалил

DmitrijEgorow commented on app/src/main/java/com/practicum/playlistmaker/search/data/TrackRepositoryImpl.kt:
```diff
@@ -1,42 +1,34 @@
+        }
+    }.flowOn(Dispatchers.IO)
 
     override fun getSearchHistory(): MutableList<Track?> {
```
```suggestion
    override fun getSearchHistory(): List<Track?> {
```

DmitrijEgorow commented on app/src/main/java/com/practicum/playlistmaker/player/ui/view_model/MediaPlayerViewModel.kt:
```diff
@@ -1,53 +1,47 @@
 import java.util.Locale
 
 class MediaPlayerViewModel(private val interactor: MediaPlayerInteractor) : ViewModel() {
     private companion object {
```
companion обычно идёт в конце, но в любом случае в рамках одного проекта стоит унифицировать логику

Подробнее описано в [документации](https://kotlinlang.org/docs/coding-conventions.html#class-layout)

DmitrijEgorow commented on app/src/main/java/com/practicum/playlistmaker/search/ui/SearchFragment.kt:
```diff
@@ -115,8 +121,6 @@ class SearchFragment : Fragment() {
-
-
         }
 
 
```
Много пустых строчек

RURURuina replied:
Вроде исправил)

DmitrijEgorow commented:
[Почему этот и другие файлы лежат в другом пакете, где-то в непонятном месте?](https://github.com/RURURuina/PlaylistMaker/blob/79aafcec0f8537e7ce6f71ebe8101897992364c7/app/src/main/java/mediateka/ui/MediatekaFragment.kt#L1)

RURURuina committed:
Completed homework 20 sprint)
Если что-то неправильно подскажите, как решить)))
Заранее спасибо за проверку
2 коммит

RURURuina commented:
Баги пофиксил

DmitrijEgorow commented:
- [ ] Всплыл новый момент, которого не было в предыдущем коммите: нужно сбрасывать в ноль время под кнопкой play/pause, когда закончилось воспроизведение композиции


https://github.com/user-attachments/assets/e7422780-5095-4a6b-8c6d-0f327f4a7a0e

RURURuina commented:
> * [ ]  Всплыл новый момент, которого не было в предыдущем коммите: нужно сбрасывать в ноль время под кнопкой play/pause, когда закончилось воспроизведение композиции
> 
>  Screen_recording_20240824_172553.mp4

Подскажите, пожалуйста, я не понимаю, какое изначально состояние таймера должно быть?
У меня сначала 0:30, после старта 00:00 и таймер по завершению нужно, чтобы было 00:00
Я спросил у однокурсников и выяснил, что у некоторых изначальный таймер 00:00 и это не ошибка
Как мне сделать, чтобы было правильно и не было крита, а то это уже 3 коммит будет, а реализация сначала 0:30, потом 00:00 звучит кривовато?

DmitrijEgorow commented:
> Я спросил у однокурсников и выяснил, что у некоторых изначальный таймер 00:00 и это не ошибка 

Сначала 00:00, затем старт с 00:00, ..., по завершению сброс в 00:00 и возврат иконки кнопки в дефолт

Так выглядит вполне нормально

RURURuina committed:
Completed homework 20 sprint)
Спасибо, что ответили на мой коммент
Таймер пофикшен
Ожидаю положительного результата))
3 коммит

RURURuina commented:
Спасибо, что ответили на мой коммент
Таймер пофикшен
Ожидаю положительного результата))
3 коммит

RURURuina commented:
Спасибо, Дмитрий)
